AsyncDisplayKit/ASMapNode.mm

```diff
@@ -0,0 +1,210 @@
+{
+  ASDN::MutexLocker l(_propertyLock);
+  _needsMapReloadOnBoundsChange = needsMapReloadOnBoundsChange;
+}
+
+- (void)fetchData
+{
+  [super fetchData];
+  if (_liveMap && !_mapNode) {
+    [self addLiveMap];
+  }
+  else {
+    [self setUpSnapshotter];
+    [self takeSnapshot];
+  }
+}
+
+- (void)clearFetchedData
+{
+  [super clearFetchedData];
+  [self removeLiveMap];
+}
+
+- (void)takeSnapshot
+{
+  if (!_snapshotter.isLoading) {
+    [_snapshotter startWithCompletionHandler:^(MKMapSnapshot *snapshot, NSError *error) {
+      if (!error) {
+        UIImage *image = snapshot.image;
+        CGRect finalImageRect = CGRectMake(0, 0, image.size.width, image.size.height);
+        
+        UIGraphicsBeginImageContextWithOptions(image.size, YES, image.scale);
+        [image drawAtPoint:CGPointMake(0, 0)];
+        
+        if (_annotations.count > 0 ) {
+          // Get a standard annotation view pin. Future implementations should use a custom annotation image property.
+          MKAnnotationView *pin = [[MKPinAnnotationView alloc] initWithAnnotation:nil reuseIdentifier:@""];
+          UIImage *pinImage = pin.image;
+          for (id<MKAnnotation>annotation in _annotations)
+          {
+            CGPoint point = [snapshot pointForCoordinate:annotation.coordinate];
+            if (CGRectContainsPoint(finalImageRect, point))
+            {
+              CGPoint pinCenterOffset = pin.centerOffset;
+              point.x -= pin.bounds.size.width / 2.0;
+              point.y -= pin.bounds.size.height / 2.0;
+              point.x += pinCenterOffset.x;
+              point.y += pinCenterOffset.y;
+              [pinImage drawAtPoint:point];
+            }
+          }
+        }
+        
+        UIImage *finalImage = UIGraphicsGetImageFromCurrentImageContext();
+        UIGraphicsEndImageContext();
+        self.image = finalImage;
+      }
+    }];
+  }
+}
+
+- (void)resetSnapshotter
+{
+  if (!_snapshotter.isLoading) {
+    _options.size = self.calculatedSize;
+    _snapshotter = [[MKMapSnapshotter alloc] initWithOptions:_options];
+  }
+}
+
+#pragma mark - Action
+- (void)addLiveMap
+{
+  if (self.isNodeLoaded && !_mapNode) {
+    _mapNode = [[ASDisplayNode alloc]initWithViewBlock:^UIView *{
+      _mapView = [[MKMapView alloc]initWithFrame:CGRectMake(0.0f, 0.0f, self.calculatedSize.width, self.calculatedSize.height)];
+      _mapView.delegate = _mapDelegate;
+      [_mapView setRegion:_options.region];
+      [_mapView addAnnotations:_annotations];
+      return _mapView;
+    }];
+    [self addSubnode:_mapNode];
+    
+    if (CLLocationCoordinate2DIsValid(_centerCoordinateOfMap)) {
+      [_mapView setCenterCoordinate:_centerCoordinateOfMap];
+    }
+  }
+}
+
+- (void)removeLiveMap
+{
+  if (_mapNode) {
+    _centerCoordinateOfMap = _mapView.centerCoordinate;
+    [_mapNode removeFromSupernode];
+    _mapView = nil;
+    _mapNode = nil;
+  }
+  self.image = nil;
```

Comment:
Is this necessary?  We may want to kill the image if some properties change like the region, although I actually think it could be better to leave it there until the new one loads.  May provide an option like clearsMapWhileLoading; if we later added this, it should be verified to work correctly with a .placeholderImage set (return to showing the placeholder if the map is cleared to load another one).

Anyway in this method let's remove self.image = nil for now.